hydromt/models/components/mesh.py

```diff
@@ -10,34 +10,389 @@
 import xarray as xr
 import xugrid as xu
 from pyproj import CRS
+from shapely.geometry import box
 
-from hydromt import workflows
 from hydromt.gis.raster import GEO_MAP_COORD
+from hydromt.io.readers import read_nc
 from hydromt.models.api import Model
+from hydromt.models.components.base import ModelComponent
+from hydromt.workflows.mesh import (
+    create_mesh2d,
+    mesh2d_from_raster_reclass,
+    mesh2d_from_rasterdataset,
+    rename_mesh,
+)
 
-__all__ = ["MeshModel"]
-logger = logging.getLogger(__name__)
+logger = getLogger(__name__)
 
 
-class MeshMixin(object):
-    # placeholders
-    # We cannot initialize an empty xu.UgridDataArray
-    _API = {
-        "mesh": Union[xu.UgridDataArray, xu.UgridDataset],
-    }
+class MeshComponent(ModelComponent):
+    """ModelComponent class for mesh components."""
 
-    def __init__(self, *args, **kwargs) -> None:
-        super().__init__(*args, **kwargs)
-        self._mesh = None
+    def __init__(self, model: Model):
+        super().__init__(model=model)
+        self._data = None
 
-    ## general setup methods
-    def setup_mesh2d_from_rasterdataset(
+    def set(
+        self,
+        data: Union[xu.UgridDataArray, xu.UgridDataset],
+        name: Optional[str] = None,
+        grid_name: Optional[str] = None,
+        overwrite_grid: bool = False,
+    ) -> None:
+        """Add data to mesh.
+
+        All layers of mesh have identical spatial coordinates in Ugrid conventions.
+        Also updates self.region if grid_name is new or overwrite_grid is True.
+
+        Parameters
+        ----------
+        data: xugrid.UgridDataArray or xugrid.UgridDataset
+            new layer to add to mesh, should contain only one grid topology.
+        name: str, optional
+            Name of new object layer, this is used to overwrite the name of
+            a UgridDataArray.
+        grid_name: str, optional
+            Name of the mesh grid to add data to. If None, inferred from data.
+            Can be used for renaming the grid.
+        overwrite_grid: bool, optional
+            If True, overwrite the grid with the same name as the grid in self.mesh.
+        """
+        # Checks on data
+        data = _check_UGrid(data, name)
+
+        # Checks on grid topology
+        # TODO: check if we support setting multiple grids at once. For now just one
+        if len(data.ugrid.grids) > 1:
+            raise ValueError(
+                "set_mesh methods only supports adding data to one grid at a time."
+            )
```

Comment:
What would that entail? I can imagine that we want one grid per component.

Reply:
I don't know. The TODO was already there in the original code and left it in for now.

Reply:
For now ``MeshComponent`` is still multi-grid. We can see later if we want to add single grid Component like ``Mesh1dComponent``, ``Mesh2dComponent`` and ``Mesh3dComponent``. The TODO can be removed and open as an issue if anyone complains. It's just that for the set methods I only allow to update data or grid schematization one grid at a time.